kernel/src/log_segment.rs

```diff
@@ -218,14 +223,109 @@ impl LogSegment {
+                                // - the checkpoint batch itself if no sidecar actions are present
+                                // - 1 or more sidecar batch that are referenced by the checkpoint batch
+                                let iterator = Self::process_checkpoint_batch(
+                                    handler.clone(),
+                                    log_root_ref.clone(),
+                                    checkpoint_batch,
+                                );
+                                match iterator {
+                                    Ok(iter) => Box::new(iter.map_ok(|batch| (batch, false))),
+                                    Err(e) => Box::new(std::iter::once(Err(e))),
+                                }
+                            } else {
+                                // Chain the checkpoint batch as is if we do not need to extract sidecar actions
+                                Box::new(std::iter::once(Ok((checkpoint_batch, false))))
+                            }
+                        } else {
+                            // Chain the checkpoint batch as is if not a UUID checkpoint batch
+                            Box::new(std::iter::once(Ok((checkpoint_batch, false))))
+                        };
+                        stream
+                    }
+                    Err(e) => {
+                        // Chain the error if any
+                        Box::new(std::iter::once(Err(e)))
+                    }
+                }
+            });
 
         Ok(commit_stream.chain(checkpoint_stream))
     }
 
+    fn process_checkpoint_batch(
+        parquet_handler: Arc<dyn ParquetHandler>,
+        log_root: Url,
+        batch: Box<dyn EngineData>,
+    ) -> DeltaResult<impl Iterator<Item = DeltaResult<Box<dyn EngineData>>> + Send> {
+        let mut visitor = SidecarVisitor::default();
+
+        // collect sidecars
+        visitor.visit_rows_of(batch.as_ref())?;
+
+        //convert sidecar actions to sidecar file paths
+        let sidecar_files: Result<Vec<_>, _> = visitor
+            .sidecars
+            .iter()
+            .map(|sidecar| Self::sidecar_to_filemeta(sidecar, &log_root))
+            .collect();
+
+        let sidecar_read_schema = get_log_add_schema().clone();
+
+        // if sidecars exist, read the sidecar files and return the iterator of sidecar actions
+        if !visitor.sidecars.is_empty() {
+            return parquet_handler.read_parquet_files(&sidecar_files?, sidecar_read_schema, None);
+        } else {
+            return Ok(Box::new(std::iter::once(Ok(batch))));
+        }
+    }
+
+    // Helper function to convert a single sidecar action to a FileMeta
+    fn sidecar_to_filemeta(sidecar: &Sidecar, log_root: &Url) -> Result<FileMeta, Error> {
+        let location = log_root
+            .join("_sidecars/")
+            .map_err(Error::from)?
+            .join(&sidecar.path)
+            .map_err(Error::from)?;
```

Comment:
I'm pretty sure `?` automatically leverages `From`?
```suggestion
        let location = log_root.join("_sidecars/")?.join(&sidecar.path)?;
```